shared/modules/metametrics.js

```diff
@@ -0,0 +1,199 @@
+ * @property {string} [metaMetricsSendCount] - number of transactions sent, used to add metametricsId
+ *  intermittently to events with onchain data attached to them used to protect identity of users.
+ */
+
+/**
+ * @typedef {Object} MetaMetricsEventPayload
+ * @property {string}  event - event name to track
+ * @property {string}  category - category to associate event to
+ * @property {boolean} [isOptIn] - happened during opt in/out workflow
+ * @property {object}  [properties] - object of custom values to track, snake_case
+ * @property {number}  [revenue] - amount of currency that event creates in revenue for MetaMask
+ * @property {string}  [currency] - ISO 4127 format currency for events with revenue, defaults to US dollars
+ * @property {number}  [value] - Abstract "value" that this event has for MetaMask.
+ * @property {boolean} [excludeMetaMetricsId] - whether to exclude the user's metametrics id for anonymity
+ * @property {MetaMetricsEventContext} [additionalContext] - additional context to attach to event
+ */
+
+/**
+ * Returns a function for tracking Segment events.
+ *
+ * @param {string} metamaskVersion - The current version of the MetaMask extension.
+ * @param {() => MetaMetricsRequiredState} getDynamicState - A function returning required fields
+ * @param {boolean} defaultMetaMetricsExclusion - Used to set the default state of exludeMetaMetricsId
+ * @returns {(payload: MetaMetricsEventPayload) => Promise<void>} - function to track an event
+ */
+export function getTrackMetaMetricsEvent (
+  metamaskVersion,
+  getDynamicState,
+  defaultMetaMetricsExclusion = false,
+) {
+  const version = process.env.METAMASK_ENVIRONMENT === 'production'
+    ? metamaskVersion
+    : `${metamaskVersion}-${process.env.METAMASK_ENVIRONMENT}`
+
+  return function trackMetaMetricsEvent ({
+    event,
+    category,
+    isOptIn,
+    properties = {},
+    revenue,
+    currency,
+    value,
+    excludeMetaMetricsId: excludeId,
+    additionalContext = {},
+  }) {
+    if (!event || !category) {
+      throw new Error('Must specify event and category.')
+    }
+    const {
+      participateInMetaMetrics,
+      context: providedContext,
+      metaMetricsId,
+      environmentType,
+      chainId,
+      network,
+      metaMetricsSendCount,
+    } = getDynamicState()
+
+    let excludeMetaMetricsId = excludeId ?? defaultMetaMetricsExclusion
+
+    // This is carried over from the old implementation, and will likely need
+    // to be updated to work with the new tracking plan. I think we should use
+    // a config setting for this instead of trying to match the event name
+    const isSendFlow = Boolean(event.match(/^send|^confirm/u))
+    if (isSendFlow && metaMetricsSendCount && !sendCountIsTrackable(metaMetricsSendCount + 1)) {
+      excludeMetaMetricsId = true
+    }
+
+    if (!participateInMetaMetrics && !isOptIn) {
+      return Promise.resolve()
+    }
+
+    const context = merge(providedContext, additionalContext, {
```

Comment:
I found it a bit confusing trying to keep track of which piece of context was passed in where. I'm not sure offhand how this can be improved, but allowing _either_ of these two values to contain _any_ part of the context object doesn't seem ideal here.

Right now if `app` or `userAgent` are provided, they'd be overwritten for example. Perhaps more importantly, the type for `providedContext` and `additionalContext` doesn't communicate that those properties are not expected to be passed in. And the same problem exists with `providedContext` vs `additionalContext`.

Maybe they can be split into two distinct types, and combined here? e.g. `MetaMetricsPageContext` and `MetaMetricsEventContext` - something like that.

Reply:
I took a stab at making this better. essentially:
the context key provided to the 'getMetaMetricsTrackEvent' can only include page or referrer. The same is true for context provided to `eventContext` (previously `additionalContext`). I also pick those keys out in case someone doesn't follow the types 😱. Locale is now a top level key of `MetaMetricsRequiredState` instead of being a subkey of context. 

let me know if this is clearer.

Reply:
Much clearer!